provider/source/src/calendar/eras.rs

```diff
@@ -0,0 +1,548 @@
+            .zip(japanese.iter().skip(2).rev())
+            .find(|(e, a)| e != a),
+        None,
+        "{japanext:?} - {japanese:?}"
+    );
+}
+
+#[test]
+fn test_calendar_eras() {
+    use icu::calendar::preferences::CalendarAlgorithm;
+    use icu::calendar::Iso;
+    use icu::calendar::{AnyCalendar, AnyCalendarKind, Date};
+    use icu::locale::extensions::unicode::Value;
+
+    let provider = crate::SourceDataProvider::new_testing();
+
+    let data = provider.all_eras().unwrap();
+
+    for (&calendar, data) in data {
+        let kind = match calendar {
+            DatagenCalendar::IslamicRgsa => continue,
+            DatagenCalendar::Gregorian => AnyCalendarKind::Gregorian,
+            DatagenCalendar::EthiopicAmeteAlem => AnyCalendarKind::EthiopianAmeteAlem,
+            DatagenCalendar::JapaneseExtended => AnyCalendarKind::JapaneseExtended,
+            DatagenCalendar::JapaneseModern => AnyCalendarKind::Japanese,
+            c => icu::calendar::AnyCalendarKind::try_from(
+                CalendarAlgorithm::try_from(&Value::try_from_str(c.cldr_name()).unwrap()).unwrap(),
+            )
+            .unwrap(),
+        };
+
+        let cal = AnyCalendar::try_new_for_kind_unstable(&provider, kind).unwrap();
+        let cal = icu::calendar::Ref(&cal);
+
+        for &(idx, ref era) in data {
+            let (in_era_iso, not_in_era_iso) = match (era.start, era.end) {
+                (Some(start), None) => {
+                    let start = Date::try_new_iso(start.year, start.month, start.day).unwrap();
+                    (start, Iso::from_fixed(Iso::to_fixed(start) - 1))
+                }
+                (None, Some(end)) => {
+                    let end = Date::try_new_iso(end.year, end.month, end.day).unwrap();
+                    (end, Iso::from_fixed(Iso::to_fixed(end) + 1))
+                }
+                _ => unreachable!(),
+            };
+
+            let in_era = in_era_iso.to_calendar(cal);
+            let not_in_era = not_in_era_iso.to_calendar(cal);
+
+            // Check that code and aliases produce identical results
+            for era in era
+                .aliases
+                .as_deref()
+                .into_iter()
+                .flat_map(|s| s.split(' '))
+                .chain(era.code.as_deref())
+            {
+                assert_eq!(
+                    Date::try_new_from_codes(
+                        Some(era),
+                        in_era.year().era_year_or_extended(),
+                        in_era.month().standard_code,
+                        in_era.day_of_month().0,
+                        cal,
+                    ),
+                    Ok(in_era)
+                );
+            }
+
+            // Unless this is the first era and it's not an inverse era, check that the
+            // not_in_era date is in a different era
+            if idx > 0 || era.end.is_some() {
+                assert_ne!(
+                    not_in_era.year().standard_era(),
```

Comment:
issue: we should also test the formatting era indices I think?

Reply:
Right, because datagen now uses the era code index in CLDR as the formatting era index.